# Creating a Simple AI Agent with LlamaIndex and OpenAI

In this tutorial, I'll guide you through building a basic AI agent using LlamaIndex and OpenAI. This approach lets you create powerful AI assistants that can use custom tools and functions to solve problems.

## What We'll Build

We'll create a simple calculator agent that can multiply numbers using OpenAI's GPT-4o-mini model and LlamaIndex's FunctionAgent framework. While this example is straightforward, the same pattern can be extended to create sophisticated agents that can:

- Search the web
- Query databases
- Call external APIs
- Process documents
- And much more

## Prerequisites

To follow along, you'll need:

1. Python installed on your machine
2. An OpenAI API key
3. Basic understanding of Python and async programming

## Setting Up the Environment

First, install the required packages:

In [6]:
#pip install llama-index python-dotenv openai -U -qqq

Create a .env file in your project directory and add your OpenAI API key:

OPENAI_API_KEY=your_api_key_here

## The Code

Let's break down our AI agent implementation:

In [7]:
import os
from dotenv import load_dotenv
import asyncio
from llama_index.core.agent.workflow import FunctionAgent
from llama_index.llms.openai import OpenAI

## Load environment variables

In [8]:
load_dotenv(override=True)
api_key = os.getenv("OPENAI_API_KEY")

# Define a simple calculator tool

In [9]:
def multiply(a: float, b: float) -> float:
    """Useful for multiplying two numbers."""
    return a * b

# Create an agent workflow with our calculator tool

In [10]:
agent = FunctionAgent(
    tools=[multiply],
    llm=OpenAI(model="gpt-4o-mini", api_key=api_key),
    system_prompt="You are a helpful assistant that can multiply two numbers."
)

## Run the agent

In [11]:
async def main():
    # Run the agent
    response = await agent.run("What is 1234 * 4567?")
    print(str(response))

# Execute the main function
response = await main()

The result of \( 1234 \times 4567 \) is 5,635,678.


## How It Works

1. **Environment Setup**: We load the OpenAI API key from our .env file.

2. **Tool Definition**: We create a simple `multiply` function that our agent can use. The docstring is important as it helps the LLM understand what the function does.

3. **Agent Creation**: We initialize a `FunctionAgent` with:
   - Our multiply tool
   - The OpenAI LLM (using the cost-effective gpt-4o-mini model)
   - A system prompt defining the agent's role

4. **Agent Execution**: We run the agent with a question asking it to multiply two numbers.

## Behind the Scenes

When you ask the agent "What is 1234 * 4567?", here's what happens:

1. The question is sent to the OpenAI model along with the tool definition
2. The model recognizes this is a multiplication problem
3. It decides to call the `multiply` function with the appropriate arguments
4. The function executes and returns the result (5,635,678)
5. The model formulates a human-friendly response with the answer

## Extending Your Agent

This simple example can be extended in numerous ways:

- Add more mathematical functions (addition, division, etc.)
- Integrate with external APIs for real-time data
- Add memory to maintain conversation context
- Implement a feedback loop for continuous improvement

## Conclusion

Building AI agents with LlamaIndex and OpenAI provides a powerful way to create AI assistants that can perform specific tasks. By defining custom tools and connecting them to large language models, you can create agents that solve complex problems in a conversational way.

The pattern shown here—defining tools, creating an agent, and running queries—forms the foundation of more complex agent architectures that can handle sophisticated workflows and real-world applications.

What will you build with your AI agent?